xla/service/gpu/fusions/custom.cc

```diff
@@ -180,14 +185,151 @@ absl::StatusOr<BufferAllocation::Slice> GetOperandSlice(
+  } else {
+    return module->entry_computation()->num_parameters() == 1 &&
+           *arg_shape ==
+               module->entry_computation()->parameter_instruction(0)->shape();
+  }
+}
+
+// Extracts the offset module as a function of the induction variable.
+std::unique_ptr<HloModule> ExtractAsFunctionOfIndvar(
+    const HloInstruction* offset_value) {
+  HloInstruction* while_op = GetWhileOp(offset_value);
+  if (while_op == nullptr) {
+    VLOG(2) << "Unable to find surrounding while op"
+            << offset_value->ToString();
+    return nullptr;
+  }
+  const HloComputation* while_body = while_op->while_body();
+  std::optional<int64_t> indvar_idx = GetLoopInductionVarTupleIdx(while_op);
+  if (indvar_idx == std::nullopt) {
+    VLOG(2) << "Unable to find induction variable index for while op: "
+            << while_op->ToString();
+    return nullptr;
+  }
+  if (while_body->num_parameters() != 1) {
+    VLOG(2) << "Expected single parameter while body: "
+            << while_body->ToString();
+    return nullptr;
+  }
+  HloInstruction* indvar = hlo_query::GetUniqueGteInstruction(
+      while_body->parameter_instruction(0), *indvar_idx);
+  if (indvar == nullptr) {
+    VLOG(2) << "Unable to find induction variable(idx=" << *indvar_idx
+            << ") in while body: " << while_body->ToString();
+    return nullptr;
+  }
+  std::unique_ptr<HloModule> offset_module = ExtractModule(
+      offset_value, -1,
+      [indvar](const HloInstruction* instr) { return instr != indvar; },
+      [](const HloInstruction*) { return ReplaceType::kReplaceParam; }, true,
+      true, true);
+  if (offset_module == nullptr) {
+    VLOG(2) << "Unable to get offset module for offset_value: "
+            << offset_value->ToString();
+    return nullptr;
+  }
+  if (!IsValid(offset_module.get(), &indvar->shape())) {
+    VLOG(2) << "Offset module does not match requirements: "
+            << offset_module->ToString();
+    return nullptr;
+  }
+  return offset_module;
+}
+
+// Extracts induction initialization module, making sure that it has zero
+// parameters.
+std::unique_ptr<HloModule> ExtractWhileInitModule(
+    const HloInstruction* while_op) {
+  std::optional<int64_t> indvar_idx = GetLoopInductionVarTupleIdx(while_op);
+  if (indvar_idx == std::nullopt) {
+    VLOG(2) << "Unable to find induction variable index for while op: "
+            << while_op->ToString();
+    return nullptr;
+  }
+  const HloInstruction* init_tuple = while_op->operand(0);
+  if (init_tuple->opcode() != HloOpcode::kTuple) {
+    VLOG(2) << "Expected init to be a tuple, found: " << init_tuple->ToString();
+    return nullptr;
+  }
+  const HloInstruction* init = init_tuple->operand(*indvar_idx);
+  std::unique_ptr<HloModule> init_module = ExtractModule(init);
+  if (init_module == nullptr) {
+    VLOG(2) << "Unable to get init module for init: " << init->ToString();
+    return nullptr;
+  }
+  if (!IsValid(init_module.get(), nullptr)) {
```

Comment:
Is it really worth factoring `IsValid`? It's pretty small, and the semantics are unclear.

Seems like you're calling it exactly twice, once with a set `arg_shape` and one with a `nullptr`. How about factoring only the part about checking for `ReplicaId` and `PartitionId`? (Although it's still not clear why you need to check those.)

Reply:
You are right. Changed it to only check for partition-id and replica-id.